File: src/core/Shared/Ticker.cs
using System;

namespace core.Shared
{
    public struct Ticker : IComparable, IEquatable<Ticker>
    {
        private readonly string _ticker;

        public Ticker(string value)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                throw new ArgumentException(nameof(value), "Ticker cannot be blank");
            }
            _ticker = value.ToUpper();
        }

        public string Value => _ticker;
        public int CompareTo(object obj)
        {
            if (obj is Ticker t)
            {
                return string.CompareOrdinal(_ticker, t._ticker);
            }

            return -1;
        }

        public override string ToString() => _ticker;

        public bool Equals(Ticker other)
        {
            return _ticker == other._ticker;
        }

        public override bool Equals(object obj)
        {
            return obj is Ticker other && Equals(other);
        }

        public override int GetHashCode()
        {
            return (_ticker != null ? _ticker.GetHashCode() : 0);
        }
    }
}

File: src/core2.fs/Shared/Ticker.fs
namespace core.Shared

open System

[<Struct>]
[<CustomEquality; CustomComparison>]
type Ticker =
    val private _ticker: string
    
    new(value: string) =
        if String.IsNullOrWhiteSpace(value) then
            invalidArg "value" "Ticker cannot be blank"
        { _ticker = value.ToUpper() }
    
    member this.Value = this._ticker
    
    override this.ToString() = this._ticker
    
    override this.Equals(obj: obj) =
        match obj with
        | :? Ticker as other -> this._ticker = other._ticker
        | _ -> false
    
    override this.GetHashCode() =
        if isNull this._ticker then 0 else this._ticker.GetHashCode()
    
    interface IComparable with
        member this.CompareTo(obj: obj) =
            match obj with
            | :? Ticker as t -> String.CompareOrdinal(this._ticker, t._ticker)
            | _ -> -1
    
    interface IEquatable<Ticker> with
        member this.Equals(other: Ticker) =
            this._ticker = other._ticker
